### ISSUE TITLE
Memory leak on m_i_gen_elem_out of DecodeTree

### ISSUE THREAD

**Issue by holio0** [Linaro/OpenCSD#55]
In 'DecodeTree::setGenTraceElemOutI' it supposed to save argument 'i_gen_trace_elem' to 'm_i_gen_elem_out', so in 'ocsd_destroy_dcd_tree' the 'm_i_gen_elem_out' can be freed via the following clean up:

```
OCSD_C_API void ocsd_destroy_dcd_tree(const dcd_tree_handle_t handle)
{
    if(handle != C_API_INVALID_TREE_HANDLE)
    {
        GenTraceElemCBObj * pIf = (GenTraceElemCBObj *)(((DecodeTree *)handle)->getGenTraceElemOutI());
        if(pIf != 0)
            delete pIf;
        ...
```

The fix can be done like this:

```
diff --git a/decoder/source/ocsd_dcd_tree.cpp b/decoder/source/ocsd_dcd_tree.cpp
index 8e29269..0158e73 100644
--- a/decoder/source/ocsd_dcd_tree.cpp
+++ b/decoder/source/ocsd_dcd_tree.cpp
@@ -174,6 +174,9 @@ void DecodeTree::setGenTraceElemOutI(ITrcGenElemIn *i_gen_trace_elem)
         pElem->getDecoderMngr()->attachOutputSink(pElem->getDecoderHandle(),i_gen_trace_elem);
         pElem = getNextElement(elemID);
     }
+    if (m_i_gen_elem_out != 0)
+        delete m_i_gen_elem_out;
+    m_i_gen_elem_out = i_gen_trace_elem;
 }
 
 ocsd_err_t DecodeTree::createMemAccMapper(memacc_mapper_t type /* = MEMACC_MAP_GLOBAL*/ )
```

**Comment by mikel-armbb** [comment]
Hi,

You are correct in that setGenTraceElemOutI() needs to store:
m_i_gen_elem_out = i_gen_trace_elem;

but the client must test and delete the object.
In this case the C-API handler code must test for a current set interface and delete a GenTraceElemCBObj  if found

Mike

**Comment by holio0** [comment]
Hi Mike,

In my usage the client is the caller of `ocsd_dt_set_gen_elem_outfn` and `ocsd_destroy_dcd_tree` defined in opencsd_c_api.h. The `ocsd_dt_set_gen_elem_outfn` allocates a `GenTraceElemCBObj` instance for function `setGenTraceElemOutI`:

```
OCSD_C_API ocsd_err_t ocsd_dt_set_gen_elem_outfn(const dcd_tree_handle_t handle, FnTraceElemIn pFn, const void *p_context)
{

    GenTraceElemCBObj * pCBObj = new (std::nothrow)GenTraceElemCBObj(pFn, p_context);
    if(pCBObj)
    {
        ((DecodeTree *)handle)->setGenTraceElemOutI(pCBObj);
```

But without assigning it to `m_i_gen_elem_out`, the `GenTraceElemCBObj` object cannot be released by `ocsd_destroy_dcd_tree`. That is, following usage can reproduce the leakage:

```
dcd_tree_handle_t dcdtree_handle;

dcdtree_handle = ocsd_create_dcd_tree(...);
ocsd_dt_set_gen_elem_outfn(dcdtree_handle, ...);
ocsd_destroy_dcd_tree(dcdtree_handle)

dcdtree_handle = ocsd_create_dcd_tree(...);
ocsd_dt_set_gen_elem_outfn(dcdtree_handle, ...);
ocsd_destroy_dcd_tree(dcdtree_handle);

```

And seems for such a caller, the opencsd_c_api.h doesn't expose a C-API to access and delete the `GenTraceElemCBObj` directly?

**Comment by mikel-armbb** [comment]
The assignment of :-
m_i_gen_elem_out = i_gen_trace_elem; 
occurs in ocsd_dcd_tree.cpp -> DecodeTree::setGenTraceElemOutI(),

A check for a previous assignment an release of the old GenTraceElemCBObj  must then take place in ocsd_c_api.cpp -> ocsd_dt_set_gen_elem_outfn() before any new object is allocated.

Thus the creation and destruction of these internal C-API layer objects takes place in the same area of code.

This will be fixed in the next release of the library

**Comment by mikel-armbb** [comment]
fixed on v1.4.1
